### ISSUE TITLE
Multi-class NMS

### ISSUE THREAD

**Issue by jenguehard** [Deci-AI/super-gradients#1319]
### 💡 Your Question

Dear all,

Can we have multi-class NMS so that we can assign only one class per object?

While detecting more than one classes, most of the times, one object is detected as class A as well as class B with same or almost similar bounding box values. Ideally, one object should be labelled as only one class. Due to similar inter-class features, such multiple labelling occur.

The option is available here under `class_agnostic_nms` : 

` def non_max_suppression(
    prediction, conf_thres=0.1, iou_thres=0.6, multi_label_per_box: bool = True, with_confidence: bool = False, class_agnostic_nms: bool = False
):
    """
    Performs Non-Maximum Suppression (NMS) on inference results

    :param prediction: raw model prediction. Should be a list of Tensors of shape (cx, cy, w, h, confidence, cls0, cls1, ...)
    :param conf_thres: below the confidence threshold - prediction are discarded
    :param iou_thres: IoU threshold for the nms algorithm
    :param multi_label_per_box: controls whether to decode multiple labels per box.
                                True - each anchor can produce multiple labels of different classes
                                       that pass confidence threshold check (default).
                                False - each anchor can produce only one label of the class with the highest score.
    :param with_confidence: whether to multiply objectness score with class score.
                            usually valid for Yolo models only.
    :param class_agnostic_nms: indicates how boxes of different classes will be treated during NMS
                               True - NMS will be performed on all classes together.
                               False - NMS will be performed on each class separately (default).
    :return: detections with shape nx6 (x1, y1, x2, y2, object_conf, class_conf, class)

    """
    candidates_above_thres = prediction[..., 4] > conf_thres  # filter by confidence
    output = [None] * prediction.shape[0]

    for image_idx, pred in enumerate(prediction):

        pred = pred[candidates_above_thres[image_idx]]  # confident

        if not pred.shape[0]:  # If none remain process next image
            continue

        if with_confidence:
            pred[:, 5:] *= pred[:, 4:5]  # multiply objectness score with class score

        box = convert_cxcywh_bbox_to_xyxy(pred[:, :4])  # cxcywh to xyxy

        # Detections matrix nx6 (xyxy, conf, cls)
        if multi_label_per_box:  # try for all good confidence classes
            i, j = (pred[:, 5:] > conf_thres).nonzero(as_tuple=False).T
            pred = torch.cat((box[i], pred[i, j + 5, None], j[:, None].float()), 1)

        else:  # best class only
            conf, j = pred[:, 5:].max(1, keepdim=True)
            pred = torch.cat((box, conf, j.float()), 1)[conf.view(-1) > conf_thres]

        if not pred.shape[0]:  # If none remain process next image
            continue

        # Apply torch batched NMS algorithm
        boxes, scores, cls_idx = pred[:, :4], pred[:, 4], pred[:, 5]
        if class_agnostic_nms:
            idx_to_keep = torchvision.ops.boxes.nms(boxes, scores, iou_thres)
        else:
            idx_to_keep = torchvision.ops.boxes.batched_nms(boxes, scores, cls_idx, iou_thres)
        output[image_idx] = pred[idx_to_keep]

    return output`

Is there a parameter in the predict function to set-up `class_agnostic_nms `? 

Best regards,

**Comment by NatanBagrov** [comment]
Hello, please see [this PR](https://github.com/Deci-AI/super-gradients/pull/1232/files) that was already merged. Are you using the latest version of SG?

**Comment by jenguehard** [comment]
Dear Natan,

Thank you for your answer. I've seen the PR, my only issue is to find the way to switch the boolean from `False` to `True` for `class_agnostic_nms` in the `model.predict()` function.

Best regards,

**Comment by NatanBagrov** [comment]
I see. At the moment this is not exposed outside, as you noticed.
You can override this method to support the flag, and I think you should also override down to `get_post_prediction_callback`

**Comment by BloodAxe** [comment]
That's correct, as of now we do not support passing `class_agnostic_nms` in predict method.
